### PR TITLE
Extend jQuery with customized show/hide/toggle

### PR DESCRIPTION
**Pull request by wxiaoguang**
Follow
* #22879

This PR makes jQuery works well in all cases, it is the first step for 
* #22847
* In the end, there will be only one hiding-method instead of the 5 old different methods, and there won't be any conflict.

### About jQuery

jQuery's `show/hide/toggle` doesn't work with our `gt-hidden` class which has `!important` 
* jQuery can not show a hidden element with `.gt-hidden { display: none !important; }` because of the `!important`.
* The `!important` is necessary for many cases. eg: overwrite the other `display: flex`

Ref: jQuery's show/hide: https://github.com/jquery/jquery/blob/main/src/css/showHide.js

### Some more info:

* The most hacky part (`.hide.show-outdate` and `.hide.hide-outdated`) has been cleaned up
* The next PR could just do a search&replace like https://github.com/go-gitea/gitea/pull/22884/commits/c6c3afcd8c593c8b495a06a7c1d6c56ab06faca7 to remove the legacy `.hide` class.
* Next PR will have more details (comments & documents) about the hiding-problem and the hiding-method.

### Screenshot (the refactored show/hide buttons work well):

<details>

#### The `show` button

image

#### Switch to the `hide` button and show details

image

</details>

### CODE REVIEW

File: web_src/less/_base.less [hunk 1]
```diff
@@ -1803,16 +1803,9 @@ footer {
...
 .hide {
   display: none;
-
-  &.show-outdated {
-    display: none !important;
-  }
-
-  &.hide-outdated {
-    display: none !important;
-  }
```
Comment (wxiaoguang) [inline web_src/less/_base.less:1815]: The most hacky part has been removed.

File: web_src/less/_review.less [hunk 2]
```diff
@@ -57,7 +57,6 @@
 .show-outdated,
 .hide-outdated {
   &:extend(.unselectable);
-  display: block !important;
```
Comment (wxiaoguang) [inline web_src/less/_review.less:60]: Another hacky part, it is also removed.

**Comment by silverwind** [comment]
Assuming we want to eventually get rid of jQuery, I would oppose to exending it.

Instead I would use simple helper functions that works with both plain `Node`/`NodeList` and optionally jQuery collections (thought ideally we would just do `$el[0]` at the call site, eliminating the need for the last `else` branch):

```js
function show(els) {
  if (els instanceof Node) {
    els.classList.remove('gt-hidden');
  } else if (els instanceof NodeList) {
    for (const el of els) {
      els.classList.remove('gt-hidden');
    }
  } else if (els instanceof $) {
    els.each(el, function() {
      this.classList.remove("gt-hidden");
    });
  }
}

function hide(els) {
  if (els instanceof Node) {
    els.classList.add('gt-hidden');
  } else if (els instanceof NodeList) {
    for (const el of els) {
      els.classList.add('gt-hidden');
    }
  } else if (els instanceof $) {
    els.each(el, function() {
      this.classList.add("gt-hidden");
    });
  }
}
```

This will allow enabling the `jquery/no-hide` and `jquery/no-show` eslint rules which is a good step on the eventual complete removal of jQuery.

File: web_src/js/jquery.js [hunk 3]
```diff
@@ -1,3 +1,76 @@
...
+  return cs ? cs[prop] : null;
+}
+
+const defaultDisplayMap = {};
+function getDefaultDisplay(el) {
+  let display = defaultDisplayMap[el.nodeName];
+  if (display) return display;
+
+  const temp = el.ownerDocument.body.appendChild(el.ownerDocument.createElement(el.nodeName));
+  display = getComputedStyleProperty(el, 'display');
```
Comment (silverwind) [inline web_src/js/jquery.js:18]: Don't `getComputedStyleProperty` in a hot code path like this, it's abhorrently slow (30ms+ in best case).

Comment (wxiaoguang) [reply web_src/js/jquery.js:18]: That's how jQuery does, and there  is a cache. So far so good, no more  problem

Comment (silverwind) [reply web_src/js/jquery.js:18]: jQuery legacy code is a very bad example to follow. Even if you cache it, I think the property accessor on the returned `style` object will still be slow regardless. But in any case, this whole solution is not something we should pursue.

Comment (wxiaoguang) [reply web_src/js/jquery.js:18]: This PR doesn't make anything worse. It's a must at the moment (to keep the old behavior, do not break anything)

You can fine  tune anything after the code gets refactored correctly.

Comment (wxiaoguang) [reply web_src/js/jquery.js:18]: I added more comments:


> to remove the calls to "getComputedStyle", it requires that there is no element with "display: none" in its own CSS classes
>
> for example: `.popup { display: none; ... } <div class="popup">` should be rewritten to `.popup { ... } <div class="popup gt-hidden">`

**Comment by wxiaoguang** [comment]
> Assuming we want to eventually get rid of jQuery, I would oppose to extending it.

This is the first step, to avoid breaking anything. You can fine tune anything after the code gets refactored correctly.

**Comment by silverwind** [comment]
What do you think about the `show`/`hide` functions above? I think they are supremely simple and universally usable. They don't rely on inline styles so work with strict CSP as well.

**Comment by wxiaoguang** [comment]
> What do you think about the `show`/`hide` functions above? I think they are supremely simple and universally usable. They don't rely on inline styles so work with strict CSP as well.

1. They do not work with `.popup { display: none; ... } <div class="popup">`, as a refactroing, keep the old logic as much as possible.
2. I do not think `el.style.display` violate CSP

**Comment by wxiaoguang** [comment]
If you want to skip this step (customize the jQuery's show/hide), then the next PR must do all the following things at the same time:
* Replace all `.hide` to `.gt-hidden`
* Replace all `$.show()/$.hide()` to the new `show()/hide()`

In my mind, "Replace all `$.show()/$.hide()` to the new `show()/hide()`" is the biggest change, it will change a lot of lines and it may need more time. So this PR just makes jQuery works with `.gt-hidden`, then there will be enough time to do more `show()/hide()` refactoring.

**Comment by wxiaoguang** [comment]
@silverwind if you still prefer to not touch the jQuery, the separate PR is
* #22916

**Comment by silverwind** [comment]
> They do not work with .popup { display: none; ... } <div class="popup">, as a refactroing, keep the old logic as much as possible.

I'd argue these should be refactored to have the hidden class initially in HTML.

> I do not think el.style.display violate CSP

It does appear `.style.xxx` is exempt from CSP as per https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src, so it should be fine indeed.  Still I prefer a universal class approach than messing directly with inline `display`.

**Comment by wxiaoguang** [comment]
> > They do not work with .popup { display: none; ... }
> 
> I'd argue these should be refactored to have the hidden class initially in HTML.

Of course, you could argue.

If you have checked every CSS style to guarantee that there is no such usage anymore, then you do not need that `getComputedStyle`

BUT, if you haven't checked, to make sure nothing would be broken, you absolutely need that `getComputedStyle` to work with `.popup { display: none; ... }`.

Could you check everything and give me a feed back?

> 
> > I do not think el.style.display violate CSP
> 
> It does appear writing to `.style.<prop>` is exempt from CSP as per [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src#violation_cases), so it should be fine indeed. Still I prefer a universal class approach than messing directly with inline `display`.

I know you prefer. I have explained my plain in #22879 , the "Solution":

> Drop all other classes/attributes, only use the .gt-hidden to hide

**Comment by wxiaoguang** [comment]
Replaced by #22950

**Pull request closed without merging**